lib/pairmotron/pair_builder.ex

```diff
@@ -0,0 +1,94 @@
+defmodule Determination do
+  defstruct dead_pairs: [], remaining_pairs: [], available_users: []
+end
+
+defmodule Pairmotron.PairBuilder do
+  @moduledoc """
+  Provides ways of determining what has changed between an existing set of %UserPair{}s and a new
+  set of %Users{}s that need to be paired.
+  %UserPair{}s are, for the intent of this module,  records of how users were paired previously.
+
+  This module will only spell out the difference  between how users have already been paired and
+  what users are now able to be paired.
+  %Pair{}s and %UserPair{}s may either not change or be deleted.
+  %Users{}s will be free to find a new pair.
+  """
+
+
+  @doc """
+  Find the dead pairs, remaining user pairs, and available users given the previous %UserPair{} records
+  and the new list of %User{}s.
+  """
+  def determify(user_pairs, users) do
+    dead_user_pairs = find_dead_user_pairs(user_pairs, users)
+    %Determination{
+      dead_pairs: dead_user_pairs |> find_dead_pairs,
+      remaining_pairs: dead_user_pairs |> find_remaining_user_pairs(user_pairs) |> unique_pairs,
+      available_users: dead_user_pairs |> find_users_to_pair(user_pairs, users)
+    }
+  end
+
+  @doc """
+  Finds the previous %UserPair{}s that will no longer be needed with the new user set.
+  """
+  def find_dead_user_pairs([], _), do: []
+  def find_dead_user_pairs(user_pairs, []), do: user_pairs
+  def find_dead_user_pairs(user_pairs, users) do
+    previously_paired_users = user_pairs |> pair_users |> MapSet.new
+    u = users |> MapSet.new
+    dead_users = MapSet.difference(previously_paired_users, u) |> MapSet.to_list
+    dead_user_pairs = dead_users
+      |> Enum.reduce([], fn(dead_user, dup) ->
+        dup ++ Enum.filter(user_pairs, fn(p) -> p.user == dead_user end)
+      end)
+    dead_matched_user_pairs = dead_user_pairs
+      |> Enum.reduce([], fn(dead_user_pair, dead_matched_user_pairs) ->
+        dead_matched_user_pairs ++ Enum.filter(user_pairs, fn(p) -> p.pair == dead_user_pair.pair && p.id != dead_user_pair.id end)
+      end)
+    dead_user_pairs ++ dead_matched_user_pairs |> Enum.uniq
+  end
+
+  @doc """
+  Finds the %Users{}s that will need to be paired because either their previous pair is no longer
+  valid or the were not previously paired.
+  """
+  def find_users_to_pair(_, []), do: []
+  def find_users_to_pair([], users), do: users
+  def find_users_to_pair(user_pairs, users) do
+    find_users_to_pair(find_dead_user_pairs(user_pairs, users), user_pairs, users)
+  end
```

Comment:
Why are these methods here? I can't seem to see where they're referenced outside of the tests.